kbs/docs/initdata.md

```diff
@@ -0,0 +1,267 @@
+user to contain any initdata in key-value format to untrusted host to inject into
+the TEE when launching. We do not limit the encoding of this data structure, which
+means that JSON and TOML are optional. This will be introduced in [Initdata Metadata](#initdata-metadata)
+- A data integrity binding mechanism. It will guide the untrusted host to bind the
+digest of the `data`` part in the Initdata Metadata to the hardware TEE-specific 
+field in evidence. This field will be checked by the verifier during the remote
+attestation. This will be introduced in 
+[Integrity Binding for Different TEEs](#integrity-binding-for-different-tees)
+- A data serialization method. This method is used to serialize the Initdata Metadata
+into a canonicalized one, which will help to get consistent cryptographic hash for
+different forms of a same Initdata Metadata. This will be introduced in
+[Data Canonicalization Algorithm](#data-canonicalization-algorithm).
+
+This spec does not define how the Initdata Metadata will be delivered into the TEE.
+Different projects will have its own way to do this. For Confidential Containers,
+we will use kata-runtime and kata-agent to collaborate to achieve this function.
+
+## Terminology
+
+This section will introduce the terminology used in this spec to avoid ambiguity.
+
+- `Initdata`: Data that needs to be injected when the TEE is started. This data
+requires integrity protection but does not require confidentiality protection.
+- `Initdata Metadata`: A data structure that includes initdata and other information
+that will help to calculate the initdata digest.
+- `Initdata digest`: Digest of the initdata. It will be used as the value of the
+TEE HOSTDATA field.
+- `HOSTDATA`: Fields that can be bound to a specific TEE instance. This field
+information will be included in the TEE-signed remote attestation report. Typically,
+Intel TDX's `mr_config_id`, AMD SNP's `hostdata` and Arm CCA's `CCA_REALM_PERSONALIZATION_VALUE`.
+In order to avoid confusion with the hostdata field of AMD SNP, when we do not
+emphasize a specific platform s.t. SNP, we are referring to the corresponding fields of
+various TEE platforms.
+
+
+## Specifications
+
+### Initdata Metadata
+
+Initdata Metadata defines a standardized initdata data format. Please note that it
+does not indicate the specific encoding format, but requires that the encoding format
+must support the expression of key-value data pairs. Typical encodings that meet
+this requirement include JSON, TOML and YAML, etc.
+
+An Initdata Metadata SHOULD have the following fields
+- `version`: The format version of the initdata metadata. Version number will provide
+extensibility. The definition in this spec is all `0.1.0`.
+- `algorithm`: The hash algorithm to calculate the value to set as `HOSTDATA`. The typical
+algorithms are `sha256`, `sha384`, `sha512`.
+- `data`: a key-value map. Including the concrete content of initdata. Nested maps are
```

Comment:
Oh. I mean
```json
{
  "data": {
    "key1": "value1",
    "object": {
      "key2": "value2"
    }
  }
}
```

Let me give an example here to make it clear

Reply:
Ahh. I might misdeliver my idea.

My idea is that the value can be another object (JSON object or TOML object), or a string. If `attestation-agent.json` is used as key, the value must be a string in CoCo scenario.

Beyond CoCo scenario, some users might want the `data` be like the following (Take JSON for example)
```
{
  "attestation-agent.json": "xxx",
  "other-settings": {
    "setting1": "yyy"
  }
}
```

How to consume the nested object is defined by the concrete use case. In CoCo, we do not use the nested format.

I am not sure if it is so flexible and I am open to changing that to a one-level key-value map.

Reply:
it looks like a special way to avoid escaping, if you decide to encode your high-level json object. in yaml you would do just:
 
 ```yaml
attestation-agent.json: xxx
other-settings: |
  {
    "setting1": "yyy"
  }
```

if it's not backed by some use case yet. I would opt to type the value as string, it could be extended later.

Reply:
In your example, `other-settings` is a string rather than a object. Try to think about this:

If some project receives initdata, but will not put them as files. It just directly use the whole `data` object. In this scenario, an object is better than a string as it could be directly deserialized.

My idea is we can support string and nested object here, but actually in CoCo we use it as string. The string equals the concrete content of the file. Yaml would be a good choice to avoid escape characters in like json.

Let me take another example such as YAML (JSON might be confused) to show the nested structure.

A one-level map of yaml initdata metadata's data field

```yaml
data:
  attestation-agent.json: |
{
  "kbs_addr": "http://172.18.0.1:8080"
}
  policy.rego: |
default CopyFileRequest := false
default CreateContainerRequest := false
```

With support of nested object
```yaml
data:
  attestation-agent.json: |
{
  "kbs_addr": "http://172.18.0.1:8080"
}
  policy.rego: |
default CopyFileRequest := false
default CreateContainerRequest := false
  some-other-information:
    key1: value1
    key2: value2
```

Let's try to image, here the `some-other-information` might not be put as files, but be directly used by something.

So you are right that if not needed, opting to type the value as string would be the best. My idea is not to rule out the potential possibility at the beginning. That is, hey it could be nested object, but it's up to you to choose whether you will use that.

Reply:
I don't think we will rule it out from the beginning if we stick to `string` payloads initially. it can be extended to allow other data types in a later revision if there is a requirement. it adds to confusion, because the object` hierarchy needs to be subject to a) json transformation and b) canonization. so other than the `string` entries, you cannot easily calculate a hash for your entry. As a general rule I think it's a good idea to keep the schema for untrusted input data as small and simple as possible.

Reply:
Ok. Sounds reasonable. During the discussion I felt it is not that convincing to keep a nested structure. Let me rewrite this section.